### PR TITLE
Fix/merkle readability and comments

### PR DESCRIPTION
**Pull request by vjdmhd**

### CODE REVIEW

File: merkle/index.go [hunk 1]
```diff
@@ -35,14 +35,19 @@ func siblingIndex(idx uint64) uint64 {
...
+// isEvenIndex returns true if the index is even
+// ex. idx = 100 then the bitwise operation idx&1 return 0
 func isEvenIndex(idx uint64) bool {
-	return idx%2 == 0
+	return idx&1 == 0
 }
 
+// getLeftIndex returns the left node index using bitwise operation
+// the left index is multiply by 2
 func getLeftIndex(idx int) int {
```
Comment (blasrodri) [inline merkle/index.go:46]: can you inline it?

File: merkle/partialtree.go [hunk 2]
```diff
@@ -20,11 +20,13 @@ func (pt *PartialTree) build(partialLayers Layers, depth uint64) (PartialTree, e
 // from merkle proof, or if a complete set of leaves provided as a first argument and no
 // helper indices given, will construct the whole tree.
 func (pt *PartialTree) buildTree(partialLayers Layers, fullTreeDepth uint64) (Layers, error) {
+	// reverse the layers to process backward
```
[the remark this answers is not in the document]
Comment (blasrodri) [reply merkle/partialtree.go:23]: on the comments it's more important to explain why, than what.

**Review by blasrodri — COMMENTED** [review]
I think this is an improvement. It would be great to have the comments focusing on why certain things are done, than what it's happening. It's easy to see the what by following the code, but sometimes why certain helper functions are called seems difficult to follow.

File: merkle/partialtree.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (vjdmhd) [reply merkle/partialtree.go:26]: completely commented in [735057e](https://github.com/ComposableFi/go-merkle-trees/pull/24/commits/735057e57a1ae409a737572b8b819b11604b5733)

File: merkle/index.go [hunk 3]
```diff
@@ -2,52 +2,83 @@ package merkle
...
+	// append all sibling indices to result
 	for i := 0; i < indicesCount; i++ {
 		siblings[i] = siblingIndex(idxs[i])
 	}
+
 	return siblings
 }
 
 // parentIndecies returns indecies of parent elements
 func parentIndecies(idxs []uint64) []uint64 {
```
Comment (blasrodri) [inline merkle/index.go:18]: can you please change `idx` to a more meaningful name? maybe something like `levelIndices`?

File: merkle/index.go [hunk 4]
```diff
@@ -2,52 +2,83 @@ package merkle
...
+
+	// loop through all indices
+	for i := 0; i < len(idxs); i++ {
+
+		// get parent index
 		parentIndex := parentIndex(idxs[i])
+
+		// checl if it is duplicate
 		isDuplicate := false
 		for j := 0; j < len(parents); j++ {
```
Comment (blasrodri) [inline merkle/index.go:30]: **Maybe** you don't need this loop at all and hence can run in O(n).
My intuition is the following:
if you have a set of indices, and want to get their parents without any duplication you can:
- get all the even indices on the level
- for the odd indices, check that you don't have a sibling that's equal to itself - 1.

Comment (blasrodri) [reply merkle/index.go:32]: @vjdmhd 

```diff
 func parentIndecies(idxs []uint64) []uint64 {
        indicesCount := len(idxs)
        var parents []uint64
+       indexFound := make(map[uint64]struct{})
        for i := 0; i < indicesCount; i++ {
                parentIndex := parentIndex(idxs[i])
-               isDuplicate := false
-               for j := 0; j < len(parents); j++ {
-                       if parentIndex == parents[j] {
-                               isDuplicate = true
-                       }
-               }
+               _, isDuplicate := indexFound[parentIndex]
                if !isDuplicate {
                        parents = append(parents, parentIndex)
+                       indexFound[parentIndex] = struct{}{}
                }
-
        }
        return parents
 }
```

However, it doesn't really improve `BenchmarkParentIndices-24`. Please double check,

Comment (blasrodri) [reply merkle/index.go:32]: Check my suggestion here: https://github.com/ComposableFi/go-merkle-trees/pull/25

**Review by blasrodri — APPROVED** [review]

**Review by thomasmodeneis — APPROVED** [review]

**Pull request merged**